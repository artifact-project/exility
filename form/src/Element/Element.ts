import Block from '@exility/block';
import css from '@exility/css';
import {ElementAttrs, ElementContext} from '../interfaces';

export default class Element extends Block<ElementAttrs, ElementContext> {
	static classNames = css({
		'input': {
			width: '100%',
			border: '1px solid red',
		},
	});

	static template = `
		var inp = context.$form[attrs.name];
	
		input.input[
			@focus
			@blur
			class.disabled=\${inp.disabled}
			class.readOnly=\${inp.readOnly}
			class.submitting=\${inp.submitting}
			name=\${inp.name}
			type=\${inp.type}
			value=\${inp.value}
			checked=\${attrs.checked}
			disabled=\${attrs.disabled}
			readOnly=\${attrs.readOnly || inp.submitting}
			autoFocus=\${attrs.autoFocus}
		]
	`;

	constructor(attrs, options) {
		super(attrs, options);
		this.getContext().register(this);
	}

	getDefaults() {
		return {type: 'text'};
	}

	getContext() {
		return this.context.$form;
	}

	disconnectedCallback() {
		this.getContext().unregister(this);
	}
}
