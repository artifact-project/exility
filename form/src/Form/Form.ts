import Block from '@exility/block';

export interface FormAttrs {
	name?: string;
	context?: FormContext;
}

export interface FormContext {
	handleEvent(event: KeyboardEvent);
}

export default class Form extends Block<FormAttrs, {[index:string]: FormContext}> {
	static template = `
		form[
			@submit="handleEvent"
			@reset="handleEvent"
			@focus="handleEvent"
			@blur="handleEvent"
			@input="handleEvent"
			@change="handleEvent"
			@keydown="handleEvent"
			@keyup="handleEvent"
			@keypress="handleEvent"
			@cut="handleEvent"
			@copy="handleEvent"
			@paste="handleEvent"
		]
			::children
	`;

	getContextForNested() {
		return {
			$form: this.getContext(),
		};
	}

	'@handleEvent'({originalEvent}) {
		this.getContext().handleEvent(originalEvent);
	}

	getContext() {
		return (
			this.attrs.context ||
			this.context[`${this.attrs.name}`] ||
			this.context.$form
		);
	}
}
