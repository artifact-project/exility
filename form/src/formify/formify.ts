import Block, {BlockClass} from '@exility/block';
import {FormContextConfig} from '../interfaces';
import {FormContext} from '../Context/Context';

export type FormifyAttrs<A, V> = A & {
	values: V;
	initialValues: V;
};

export default function formify<V extends object = {}, A = {}>(config: FormContextConfig<V>) {
	return (Target: BlockClass<A>): Block<FormifyAttrs<A, V>> => {
		return <any>class Formify extends Block<FormifyAttrs<A, V>> {
			private formContext: FormContext<V>;

			static blocks = {FormBlock: Target};
			static template = `FormBlock[__attrs__=\${attrs}]`;

			constructor(attrs: FormifyAttrs<A, V>, options) {
				super(attrs, options);
			}

			getContextForNested() {
				if (this.formContext === void 0) {
					this.formContext = new FormContext<V>(this.attrs.initialValues, config);
				}

				return {$form: this.formContext};
			}

			attributeChangedCallback(name: string, newValue) {
				if (name === 'initialValues') {
					this.formContext.setInitialValues(newValue);
				}
			}

			getDefaults() {
				return {initialValues: {}};
			}
		};
	};
};
