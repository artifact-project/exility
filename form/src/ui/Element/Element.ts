import Block from '@exility/block';
import {IFormElement, UIElementAttrs, UIElementContext} from '../../interfaces';

export default class UIElement extends Block<UIElementAttrs, UIElementContext> {
	static classNames = true;

	static template = `
		const elem = __this__.getFormElement();
		const tag = (elem.type === 'textarea' || elem.type === 'select') ? elem.type : 'input';
	
		trait inputtable > .is-\${elem.type}[
			ref="input"
		
			@blur
			@focus
			
			@input="input"
			@change="input"
			
			@keydown="input"
			@keyup="input"
			@keypress="input"
			
			@cut="input"
			@copy="input"
			@paste="input"
						
			class.shape-\${elem.shape}=\${elem.shape}
			class.size-\${elem.size}=\${elem.size}
			
			class.active=\${elem.active}
			class.checked=\${elem.checked}
			class.readOnly=\${elem.readOnly}
			class.disabled=\${elem.disabled}
			class.changed=\${elem.changed}
			class.touched=\${elem.touched}
			class.invalid=\${elem.invalid}
			class.submitting=\${elem.submitting}
			
			name=\${elem.name}
			type=\${elem.type}
			value=\${elem.value}
			checked=\${elem.checked}
			disabled=\${elem.disabled}
			readOnly=\${elem.readOnly || elem.submitting}
			autoFocus=\${elem.autoFocus}
			maxLength=\${elem.maxLength}
			selectedIndex=\${elem.selectedIndex}
			placeholder=\${elem.placeholder}
		]
		
		if (tag === 'select')
			select[+inputtable]
				for (opt in attrs.options)
					option[value=\${opt.value}] | \${opt.text}
		else
			\${tag}[+inputtable]
	`;

	private input: HTMLInputElement;
	private formElement: IFormElement;

	registerRef(name: 'input', el) {
		this[name] = el;
	}

	getFormElement() {
		if (this.formElement == null) {
			this.formElement = this.context.$form.connectElement(this);
		}

		return this.formElement;
	}

	connectedCallback() {
		this.getFormElement();
	}

	disconnectedCallback() {
		this.context.$form.disconnectElement(this);
		this.input = null;
		this.formElement = null;
	}

	focus() {
		setTimeout(() => {
			try {
				this.input.focus();
			} catch (err) {
				console.warn('[@exility/form] ui/Element', err);
			}
		});
	}

	getDefaults() {
		return {
			type: 'text',
		};
	}
}
