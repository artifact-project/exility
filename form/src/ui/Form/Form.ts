import Block from '@exility/block';
import {UIFormAttrs, UIFormContext} from '../../interfaces';
import UIElement from '../Element/Element';

export default class UIForm extends Block<UIFormAttrs, UIFormContext> {
	static blocks = {
		Element: UIElement,
	};

	static template = `
		const form = context.$form;
	
		form.\${form.id}[
			@submit="handleEvent"
			@reset="handleEvent"
			
			@blur="input"
			@focus="input"
			
			class.changed=\${form.changed}
			class.invalid=\${form.invalid}
			class.locked=\${form.locked}
			
			class.submitting=\${form.submitting}
			class.submitFailed=\${form.submitFailed}
			class.submitSucceeded=\${form.submitSucceeded}
		]
			::children
	`;

	connectedCallback() {
		this.context.$form.connectForm(this);
	}

	disconnectedCallback() {
		this.context.$form.disconnectForm(this);
	}

	'@input'({target, originalEvent}) {
		console.log(arguments[0]);
		this.context.$form.handleEvent(target, originalEvent);
	}
}
