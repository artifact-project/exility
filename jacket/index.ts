import mount, {DOMWrapper} from './src/mount/mount';

export {
	mount,
	DOMWrapper,
};
