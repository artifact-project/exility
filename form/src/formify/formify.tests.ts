import Block from '@exility/block';
import {create} from '@exility/jacket';
import {requestFrame} from '@perf-tools/balancer';

// UI
import Form from '../ui/Form/Form';
import Element from '../ui/Element/Element';
import formify from './formify';

async function frame() {
	return new Promise(resolve => {
		requestFrame(resolve);
	});
}

@formify<{login: string}, {}>({
	submit: () => new Promise(resolve => {
		setTimeout(resolve, 20);
	}),
})
class Login extends Block<{}, UIFormContext> {
	static blocks = {Form, Element};
	static template = `
		Form
			Element[name="email" required]
			button[type="submit"] | Send
	`;
}

it('formify', async () => {
	const wrapper = create(Login);

	await frame();
	expect(wrapper).toMatchSnapshot();

	wrapper.find('input').val('ibn@rubaxa.org').simulate('input');
	await frame();
	expect(wrapper).toMatchSnapshot();

	wrapper.find('button').click();
	await frame();
	expect(wrapper.classList).toEqual(['changed', 'submitting']);
	expect(wrapper.find('input').classList).toEqual(['changed', 'is-text', 'submitting']);
	expect(wrapper.find('input').attr('readOnly')).toBe(true);

	await frame();
	expect(wrapper.classList).toEqual(['submitSucceeded']);
	expect(wrapper.find('input').classList).toEqual(['is-text']);
	expect(wrapper.find('input').attr('readOnly')).toBe(false);
});
