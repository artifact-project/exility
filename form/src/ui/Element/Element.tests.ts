import {create} from '@exility/jacket';
import UIElement from './Element';
import {FormContext} from '../../Context/Context';
import {requestFrame} from '@perf-tools/balancer';
import {UIElementAttrs, UIElementContext} from '../../interfaces';

describe('ui / Element', () => {
	let context;

	async function frame() {
		return new Promise(resolve => {
			requestFrame(resolve);
		});
	}

	async function simulateEvent(wrapper, event = 'input') {
		context.$form.handleEvent(wrapper.target, {
			type: event,
			target: wrapper.target.getRootNode(),
		});
		await frame();
	}

	beforeEach(() => {
		context = {
			$form: new FormContext({
				inline: 'Wow!',
				array: ['foo'],
				switcher: 'bar',
				color: 'blue',
			}, {
				rules: {
					// Only English alphabet
					username: ({value}) => /^[a-z]*$/i.test(value) ? null : {
						id: 'onlyEng',
						detail: null,
					},
				},
				submit: () => Promise.resolve(true),
			}),
		};
	});

	describe('text', () => {
		it('empty', async () => {
			const text = create<UIElementAttrs, UIElementContext>(
				UIElement,
				{name: 'empty'},
				context,
			);

			expect(text).toMatchSnapshot();
			expect((text.target as any).input instanceof HTMLInputElement).toBe(true);
			expect(context.$form.values).toEqual({empty: ''});

			text.attr('value', 'ok');
			await simulateEvent(text);

			expect(context.$form.values).toEqual({empty: 'ok'});
		});

		it('with initial value', async () => {
			const text = create(
				UIElement,
				{name: 'empty', value: 'Foo'},
				context,
			);
			expect(text).toMatchSnapshot();
			expect(context.$form.values).toEqual({empty: 'Foo'});
		});

		it('with value', async () => {
			expect(create<UIElementAttrs, UIElementContext>(
				UIElement,
				{name: 'inline'},
				context,
			)).toMatchSnapshot();
		});
	});

	describe('checkbox', () => {
		it('empty', async () => {
			const cbx = create(
				UIElement,
				{name: 'empty', type: 'checkbox'},
				context,
			);

			expect(cbx).toMatchSnapshot();
			expect(cbx.attr('checked')).toBe(false);
		});

		it('with inline checked', async () => {
			const cbx = create(
				UIElement,
				{name: 'empty', type: 'checkbox', checked: true},
				context,
			);

			expect(cbx.attr('checked')).toBe(true);
		});

		it('with checked: foo', async () => {
			const cbx = create(
				UIElement,
				{name: 'array', type: 'checkbox', value: 'foo'},
				context,
			);

			expect(cbx).toMatchSnapshot();
			expect(cbx.attr('checked')).toBe(true);
		});

		it('with checked: bar', async () => {
			const foo = create(
				UIElement,
				{name: 'array', type: 'checkbox', value: 'foo'},
				context,
			);

			const bar = create(
				UIElement,
				{name: 'array', type: 'checkbox', value: 'bar'},
				context,
			);

			expect(bar).toMatchSnapshot();
			expect(context.$form.values).toEqual({array: ['foo']});
			expect(foo.attr('checked')).toBe(true);
			expect(bar.attr('checked')).toBe(false);

			bar.attr('checked', true);
			await simulateEvent(bar);
			expect(context.$form.values).toEqual({array: ['foo', 'bar']});

			foo.attr('checked', false);
			await simulateEvent(foo);
			expect(context.$form.values).toEqual({array: ['bar']});

			bar.attr('checked', false);
			await simulateEvent(bar);
			expect(context.$form.values).toEqual({array: []});
		});
	});

	describe('radio', () => {
		it('empty', async () => {
			const radio = create(
				UIElement,
				{name: 'empty', type: 'radio'},
				context,
			);

			expect(radio).toMatchSnapshot();
			expect(radio.attr('checked')).toBe(false);
		});

		it('with checked', async () => {
			const foo = create(
				UIElement,
				{name: 'switcher', type: 'radio', value: 'foo'},
				context,
			);

			const bar = create(
				UIElement,
				{name: 'switcher', type: 'radio', value: 'bar'},
				context,
			);

			const baz = create(
				UIElement,
				{name: 'switcher', type: 'radio', value: 'baz'},
				context,
			);

			expect(foo.attr('checked')).toBe(false);
			expect(bar.attr('checked')).toBe(true);
			expect(baz.attr('checked')).toBe(false);
			expect(context.$form.values).toEqual({switcher: 'bar'});

			baz.attr('checked', true);
			await simulateEvent(baz);

			expect(foo.attr('checked')).toBe(false);
			expect(bar.attr('checked')).toBe(false);
			expect(baz.attr('checked')).toBe(true);
			expect(context.$form.values).toEqual({switcher: 'baz'});
		});
	});

	describe('textarea', () => {
		it('with initial value', async () => {
			expect(create(
				UIElement,
				{name: 'empty', value: 'Foo', type: 'textarea'},
				context,
			)).toMatchSnapshot();
		});
	});

	describe('select', () => {
		it('with value', async () => {
			const select = create(
				UIElement,
				{
					name: 'color',
					type: 'select',
					options: [
						{value: 'red', text: 'Red'},
						{value: 'green', text: 'Green'},
						{value: 'blue', text: 'Blue'},
					],
				},
				context,
			);

			expect(select).toMatchSnapshot();
			expect(select.attr('selectedIndex')).toBe(2);
		});
	});

	it('focus/blur', async () => {
		const text = create<UIElementAttrs, UIElementContext>(
			UIElement,
			{name: 'empty'},
			context,
		);

		await simulateEvent(text, 'focus');
		expect(text.classList).toEqual(['active', 'is-text']);

		await simulateEvent(text, 'blur');
		expect(text.classList).toEqual(['is-text', 'touched']);
	});

	it('changed', async () => {
		const text = create<UIElementAttrs, UIElementContext>(
			UIElement,
			{name: 'empty'},
			context,
		);

		await simulateEvent(text, 'focus');
		text.attr('value', 'wow');
		await simulateEvent(text);

		expect(text.classList).toEqual(['active', 'changed', 'is-text']);

		await simulateEvent(text, 'blur');
		expect(text.classList).toEqual(['changed', 'is-text', 'touched']);

		text.attr('value', '');
		await simulateEvent(text);
		expect(text.classList).toEqual(['is-text', 'touched']);
	});

	it('invalid / required', async () => {
		const text = create(
			UIElement,
			{name: 'username', required: true},
			context,
		);
		const textFormElem = (text.target as UIElement).getFormElement();

		await frame();
		expect(!!textFormElem.errors).toBe(true);
		expect(text.classList).toEqual(['invalid', 'is-text']);

		text.attr('value', 'X');
		await simulateEvent(text);
		expect(text.classList).toEqual(['changed', 'is-text']);
		expect(textFormElem.errors).toEqual({});
	});

	it('invalid / minLength', async () => {
		const text = create(
			UIElement,
			{name: 'username', minLength: 3},
			context,
		);
		const textFormElem = (text.target as UIElement).getFormElement();

		text.attr('value', 'X');
		await simulateEvent(text);

		expect(!!textFormElem.errors.minLength).toBe(true);
		expect(text.classList).toEqual(['changed', 'invalid', 'is-text']);

		text.attr('value', 'Xy!');
		await simulateEvent(text);
		expect(!!textFormElem.errors.minLength).toBe(false);
		expect(!!textFormElem.errors.onlyEng).toBe(true);
		expect(text.classList).toEqual(['changed', 'invalid', 'is-text']);

		text.attr('value', 'Xyz');
		await simulateEvent(text);
		expect(text.classList).toEqual(['changed', 'is-text']);
		expect(textFormElem.errors).toEqual({});
	});

	it('readOnly/LockedForm', async () => {
		const text = create<UIElementAttrs, UIElementContext>(
			UIElement,
			{name: 'empty'},
			context,
		);

		context.$form.lock();
		expect(text.attr('readOnly')).toBe(true);
		expect(text.classList).toEqual(['is-text', 'readOnly']);

		context.$form.unlock();
		expect(text.attr('readOnly')).toBe(false);
		expect(text.classList).toEqual(['is-text']);
	});
});
