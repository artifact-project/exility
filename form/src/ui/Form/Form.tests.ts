import {create} from '@exility/jacket';
import Block from '@exility/block';
import Form from './Form';
import Element from '../Element/Element';
import {FormContext} from '../../Context/Context';
import {requestFrame} from '@perf-tools/balancer';
import {UIFormContext} from '../../interfaces';

describe('ui / Form', () => {
	let context;

	async function frame() {
		return new Promise(resolve => {
			requestFrame(resolve);
		});
	}

	beforeEach(() => {
		context = {
			$form: new FormContext({
				inline: 'Wow!',
				array: ['foo'],
				switcher: 'bar',
				color: 'blue',
			}, {
				rules: {
					// Only English alphabet
					username: ({value}) => /^[a-z]*$/i.test(value) ? null : {
						id: 'onlyEng',
						detail: null,
					},
				},
				submit: () => Promise.resolve(true),
			}),
		};
	});

	function createForm(children) {
		class MyForm extends Block<{}, UIFormContext> {
			static blocks = {Form, Element};
			static template = `Form > ${children}`;
		}

		return create(
			MyForm,
			{},
			context,
		);
	}

	test('changed / empty', async () => {
		const hi = createForm(`Element[name="empty"]`);

		await frame();
		expect(hi).toMatchSnapshot();
		expect(hi.classList).toEqual([]);

		hi.find('input').val('123098').simulate('input');

		await frame();
		expect(hi).toMatchSnapshot();
		expect(hi.classList).toEqual(['changed']);
	});

	test('changed / inline', async () => {
		const hi = createForm(`Element[name="inline" minLength="6"]`);

		await frame();
		expect(hi).toMatchSnapshot();
		expect(hi.classList).toEqual(['invalid']);

		hi.find('input').val('Exility/Value').simulate('input');

		await frame();
		expect(hi).toMatchSnapshot();
		expect(hi.classList).toEqual(['changed']);
	});
});
