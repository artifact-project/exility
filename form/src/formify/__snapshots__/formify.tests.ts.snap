// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formify 1`] = `
<form class="invalid">
  <input class="invalid is-text"
         name="email"
         type="text"
         value
  >
  <button type="submit">
    Send
  </button>
</form>
`;

exports[`formify 2`] = `
<form class="changed">
  <input class="changed is-text"
         name="email"
         type="text"
         value="ibn@rubaxa.org"
  >
  <button type="submit">
    Send
  </button>
</form>
`;
